File: FInd_Peak_Element.cpp
/*162. Find Peak Element  QuestionEditorial Solution  My Submissions
Total Accepted: 84303 Total Submissions: 241583 Difficulty: Medium
A peak element is an element that is greater than its neighbors.

Given an input array where num[i] ≠ num[i+1], find a peak element and return its index.

The array may contain multiple peaks, in that case return the index to any one of the peaks is fine.

You may imagine that num[-1] = num[n] = -∞.

For example, in array [1, 2, 3, 1], 3 is a peak element and your function should return the index number 2.

click to show spoilers.
Note:
Your solution should be in logarithmic complexity.

Credits:
Special thanks to @ts for adding this problem and creating all test cases.

Hide Company Tags Microsoft Google
Hide Tags Binary Search Array
*/



/*binary search O(logn) time
since we only need to return the index of anyone of the peaks
If num[i-1] < num[i] > num[i+1], then num[i] is peak
If num[i-1] < num[i] < num[i+1], then num[i+1...n-1] must contains a peak
If num[i-1] > num[i] > num[i+1], then num[0...i-1] must contains a peak
If num[i-1] > num[i] < num[i+1], then both sides have peak
(n is num.length)
*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        return helper(nums, 0, nums.size()-1);
    }
    
    int helper(vector<int>& nums, int start, int end){
        if(start==end){
            return start;
        } else if (start+1==end){
            if(nums[start] > nums[end]) return start;
            return end;
        } else {
            int mid = (start+end)/2;
            if(nums[mid-1]<nums[mid] && nums[mid]>nums[mid+1]){
                return mid;
            } else if(nums[mid-1]<nums[mid] && nums[mid]<nums[mid+1]){
                return helper(nums, mid+1, end);
            } else {
                return helper(nums, start, mid-1);
            }
        }
    }
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int findPeakElement(int[] nums) {
        return helper(nums, 0, nums.length-1);
    }
    
    public int helper(int[] nums, int start, int end){
        if(start==end){
            return start;
        } else if (start+1 == end){
            if(nums[start] > nums[end]) return start;
            return end;
        } else {
            int mid = (start+end)/2;
            if(nums[mid-1]<nums[mid] && nums[mid]>nums[mid+1]){
                return mid;
            } else if(nums[mid-1]<nums[mid] && nums[mid]<nums[mid+1]){
                return helper(nums, mid+1,end);
            } else {
                return helper(nums, start, mid-1);
            }
        }
    }
}

File: Find_Peak_Element.swift
/*
A peak element is an element that is strictly greater than its neighbors.

Given an integer array nums, find a peak element, and return its index. If the array contains multiple peaks, return the index to any of the peaks.

You may imagine that nums[-1] = nums[n] = -∞.



Example 1:

Input: nums = [1,2,3,1]
Output: 2
Explanation: 3 is a peak element and your function should return the index number 2.
Example 2:

Input: nums = [1,2,1,3,5,6,4]
Output: 5
Explanation: Your function can return either index number 1 where the peak element is 2, or index number 5 where the peak element is 6.


Constraints:

1 <= nums.length <= 1000
-231 <= nums[i] <= 231 - 1
nums[i] != nums[i + 1] for all valid i.


Follow up: Could you implement a solution with logarithmic complexity?
*/

/*
Solution 2:
binary search

use left+1 < right
- right = mid
- left = mid
- return nums[left] > nums[right] ? left : right

Time Complexity: O(log n)
Space Complexity: O(1)
*/
class Solution {
    func findPeakElement(_ nums: [Int]) -> Int {
        if nums.count <= 1 { return 0 }
        var left = 0
        var right = nums.count - 1
        while left+1 < right {
            let mid = left + (right-left)/2
            if nums[mid] > nums[mid+1] {
                right = mid
            } else {
                left = mid
            }
        }

        return nums[left] > nums[right] ? left : right
    }
}

/*
Solution 1:
Binary search

use left < right
- right = mid
- left = mid+1
- return left

Time Complexity: O(log n)
Space Complexity: O(1)
*/
class Solution {
    func findPeakElement(_ nums: [Int]) -> Int {
        if nums.count <= 1 { return 0 }
        var left = 0
        var right = nums.count - 1
        while left < right {
            let mid = left + (right-left)/2
			// compare mid & mid+1 here
            if nums[mid] > nums[mid+1] {
                right = mid
            } else {
                left = mid+1
            }
        }

        return left
    }
}

/*
one more binary search solution writing way
*/
class Solution {
    func findPeakElement(_ nums: [Int]) -> Int {
        let n = nums.count

        var left = 0
        var right = n-1
        while left <= right {
            let mid = left + (right-left)/2
            if ( mid == 0
                || (mid > 0 && nums[mid] > nums[mid-1]))
            && ( mid == n-1
                || (mid+1 < n && nums[mid] > nums[mid+1])) {
                return mid
            }

            if mid > 0 && nums[mid] < nums[mid-1] {
                right = mid-1
            } else if mid+1 < n && nums[mid] < nums[mid+1] {
                left = mid+1
            }
        }
        return -1
    }
}